File: src/algorithms/strings/group-anagrams/solution2.js
// O(w * n * log(n)) time | O(wn) space
// w = # of words
// n = length of the longest word
function groupAnagrams(words) {
	const anagrams = {};
	for (const word of words) {
		const sortedWord = word.split('').sort().join('');
		if (sortedWord in anagrams) {
			anagrams[sortedWord].push(word);
		} else {
			anagrams[sortedWord] = [word];
		}
	}
	return Object.values(anagrams);
}

exports.groupAnagrams = groupAnagrams;

File: src/algorithms/strings/group-anagrams/solution2.go
package main

import "sort"

// O(w * n * log(n)) time | O(wn) space 
// w = number of words, n = length of the longest word
func GroupAnagrams(words []string) [][]string {
	anagrams := map[string][]string{}

	for _, word := range words {
		sortedWord := sortWord(word)
		anagrams[sortedWord] = append(anagrams[sortedWord], word)
	}

	result := [][]string{}
	for _, group := range anagrams {
		result = append(result, group)
	}
	return result
}

func sortWord(word string) string {
	wordBytes := []byte(word)
	sort.Slice(wordBytes, func(i, j int) bool {
		return wordBytes[i] < wordBytes[j]
	})
	return string(wordBytes)
}